File: seeds/pythonboard/src/traversal/machine.py
from .traversal_types import Edge, GraphDescriptor, TraversalResult, NodeDescriptor
from typing import Awaitable, List, Optional, Self
from .state import MachineEdgeState
from .representation import GraphRepresentation
from .result import MachineResult
from .iterator import TraversalMachineIterator

class TraversalMachine():
  graph: GraphRepresentation
  previousResult: Optional[TraversalResult]

  def __init__(self, descriptor: GraphDescriptor, result: Optional[TraversalResult] = None):
    self.graph = GraphRepresentation(descriptor)
    self.previousResult = result

  def __aiter__(self):
    return self.start()

  def start(self):
    if self.previousResult:
      return TraversalMachineIterator(self.graph, self.previousResult)

    entries = self.graph.entries
    if len(entries) == 0:
        raise Exception("No entry node found in graph.")
    # Create fake edges to represent entry points.
    opportunities = [Edge(previous="$entry", next=entry) for entry in entries]
    entryResult = MachineResult(
      NodeDescriptor(id="$empty", type="$empty"),
      {},
      [],
      opportunities,
      [],
      MachineEdgeState(),
      {},
    )
    return TraversalMachineIterator(self.graph, entryResult)

  @staticmethod
  async def prepareToSafe(result: TraversalResult) -> Awaitable[TraversalResult]:
    return await TraversalMachineIterator.processAllPendingNodes(result)

File: packages/breadboard/src/traversal/machine.ts
/**
 * @license
 * Copyright 2023 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { StartLabel } from "@breadboard-ai/types";
import type { GraphDescriptor, TraversalResult } from "../types.js";
import { TraversalMachineIterator } from "./iterator.js";
import { GraphRepresentation } from "./representation.js";
import { MachineResult } from "./result.js";
import { MachineEdgeState } from "./state.js";

export class TraversalMachine implements AsyncIterable<TraversalResult> {
  graph: GraphRepresentation;
  previousResult?: TraversalResult;

  constructor(
    descriptor: GraphDescriptor,
    result?: TraversalResult,
    start?: StartLabel
  ) {
    this.graph = new GraphRepresentation(descriptor, start);
    this.previousResult = result;
  }

  [Symbol.asyncIterator](): AsyncIterator<TraversalResult> {
    return this.start();
  }

  start(): TraversalMachineIterator {
    if (this.previousResult)
      return new TraversalMachineIterator(this.graph, this.previousResult);

    const { entries } = this.graph;
    if (entries.length === 0) throw new Error("No entry node found in graph.");
    // Create fake edges to represent entry points.
    const opportunities = entries.map((entry) => ({
      from: "$entry",
      to: entry,
    }));
    const entryResult = new MachineResult(
      { id: "$empty", type: "$empty" },
      {},
      [],
      { from: "$entry", to: entries[0] },
      opportunities,
      [],
      new MachineEdgeState()
    );
    return new TraversalMachineIterator(this.graph, entryResult);
  }
}
